{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7fb27b941602401d91542211134fc71a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import os\n",
    "import datetime\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pyproj import CRS, Transformer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c9f9bdc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ll2xy(lon, lat):\n",
    "    \"\"\"\n",
    "    Transform coordinates from input geodetic coordinates (lon, lat)\n",
    "    to output Antarctic Polar Stereographic coordinates (x, y)\n",
    "\n",
    "    Parameters\n",
    "    lon - Geodetic longitude in EPSG:4326 [float]\n",
    "    lat - Geodetic latitude in EPSG:4326 [float]\n",
    "\n",
    "    Returns\n",
    "    x - Antarctic Polar Stereographic (EPSG:3031) x [float]\n",
    "    y - Antarctic Polar Stereographic (EPSG:3031) y [float]\n",
    "    \"\"\"\n",
    "\n",
    "    crs_ll = CRS(\"EPSG:4326\")\n",
    "    crs_xy = CRS(\"EPSG:3031\")\n",
    "    ll_to_xy = Transformer.from_crs(crs_ll, crs_xy, always_xy=True)\n",
    "    x, y = ll_to_xy.transform(lon, lat)\n",
    "    return x, y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "493c7b0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load(file):\n",
    "    \"\"\"\n",
    "    Load processed gps data file into pandas table.\n",
    "\n",
    "    Parameters\n",
    "    file - .pos precise point solution file, Natural Resources Canada [string]\n",
    "\n",
    "    Returns\n",
    "    data - Pandas DataFrame with the following columns, extracted from file\n",
    "        time - Time as datetime object\n",
    "        day_of_year - Time as julian day\n",
    "    flip - flag to flip order of readings for 2024 csrs-ppp data\n",
    "    \"\"\"\n",
    "\n",
    "    data = pd.DataFrame()  # Create Pandas DataFrame\n",
    "    flip = False\n",
    "    # Read data file\n",
    "\n",
    "    # Convert longitude and latitude from deg min sec to fractional degrees\n",
    "    # Three different file formats so try one first and try the other if it\n",
    "    # throws a not found exception.\n",
    "\n",
    "    try:\n",
    "        # CSRS-PPP 2024\n",
    "        d = pd.read_csv(file, skiprows=3, sep=\"\\s+\")\n",
    "        data[\"longitude\"] = d[\"LONDD\"] - d[\"LONMN\"] / 60 - d[\"LONSS\"] / 60 / 60\n",
    "        data[\"latitude\"] = d[\"LATDD\"] - d[\"LATMN\"] / 60 - d[\"LATSS\"] / 60 / 60\n",
    "        data[\"time\"] = pd.to_datetime(d[\"YEAR-MM-DD\"] + \"T\" + d[\"HR:MN:SS.SS\"])\n",
    "        data[\"day_of_year\"] = d[\"DAYofYEAR\"]\n",
    "\n",
    "        # Look at data and decide if to flip\n",
    "        if len(data.index) > 1:\n",
    "            diff = data[\"time\"].iloc[0] - data[\"time\"].iloc[1]\n",
    "            if diff > datetime.timedelta(seconds=0):\n",
    "                flip = True\n",
    "    except:  # noqa: E722\n",
    "        try:\n",
    "            d = pd.read_csv(file, skiprows=7, delim_whitespace=True)\n",
    "            data[\"longitude\"] = d[\"LON(d)\"] - d[\"LON(m)\"] / 60 - d[\"LON(s)\"] / 60 / 60\n",
    "            data[\"latitude\"] = d[\"LAT(d)\"] - d[\"LAT(m)\"] / 60 - d[\"LAT(s)\"] / 60 / 60\n",
    "            data[\"time\"] = pd.to_datetime(d[\"YEAR-MM-DD\"] + \"T\" + d[\"HR:MN:SS.SSS\"])\n",
    "            data[\"day_of_year\"] = d[\"DOY\"]\n",
    "        except:  # noqa: E722\n",
    "            try:\n",
    "                d = pd.read_csv(file, skiprows=5, delim_whitespace=True)\n",
    "                data[\"longitude\"] = d[\"LONDD\"] - d[\"LONMN\"] / 60 - d[\"LONSS\"] / 60 / 60\n",
    "                data[\"latitude\"] = d[\"LATDD\"] - d[\"LATMN\"] / 60 - d[\"LATSS\"] / 60 / 60\n",
    "                data[\"time\"] = pd.to_datetime(d[\"YEAR-MM-DD\"] + \"T\" + d[\"HR:MN:SS.SS\"])\n",
    "                data[\"day_of_year\"] = d[\"DAYofYEAR\"]\n",
    "            except:  # noqa: E722\n",
    "                d = pd.read_csv(file, skiprows=6, delim_whitespace=True)\n",
    "                data[\"longitude\"] = (\n",
    "                    d[\"LON(d)\"] - d[\"LON(m)\"] / 60 - d[\"LON(s)\"] / 60 / 60\n",
    "                )\n",
    "                data[\"latitude\"] = (\n",
    "                    d[\"LAT(d)\"] - d[\"LAT(m)\"] / 60 - d[\"LAT(s)\"] / 60 / 60\n",
    "                )\n",
    "                data[\"time\"] = pd.to_datetime(d[\"YEAR-MM-DD\"] + \"T\" + d[\"HR:MN:SS.SSS\"])\n",
    "                data[\"day_of_year\"] = d[\"DOY\"]\n",
    "\n",
    "    x, y = ll2xy(lon=data[\"longitude\"], lat=data[\"latitude\"])\n",
    "    data[\"x\"] = x\n",
    "    data[\"y\"] = y\n",
    "\n",
    "    x0 = data[\"x\"][0]\n",
    "    y0 = data[\"y\"][0]\n",
    "\n",
    "    data[\"dist\"] = np.sqrt((data[\"x\"] - x0) ** 2 + (data[\"y\"] - y0) ** 2)\n",
    "    return data, flip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "03a59138",
   "metadata": {},
   "outputs": [],
   "source": [
    "def datastream(dir, years):\n",
    "    \"\"\"\n",
    "    Takes input directory dir with year subdirectories containign .pos files.\n",
    "    Starting from the first file in the first year, append days until all files\n",
    "    in the directory have been added.\n",
    "\n",
    "    Parameters\n",
    "    dir - Input directory tree with structure described in top comment [string]\n",
    "    years - years to be run [arr of strings]\n",
    "\n",
    "    Returns\n",
    "    data - Resulting multiday time series, data. [DataFrame]\n",
    "    \"\"\"\n",
    "    data = pd.DataFrame()\n",
    "    for year in years:\n",
    "        for folder in os.scandir(dir.path):\n",
    "            if folder.is_dir():  # folder is an os.DirEntry object\n",
    "                if folder.name == year:\n",
    "                    print(folder.name)\n",
    "                    for gps in os.listdir(folder.path):\n",
    "                        if gps.endswith(\".pos\") and not gps.startswith(\n",
    "                            \".\"\n",
    "                        ):  # Ignore xyzt, zip files\n",
    "                            ind_data, flip = load(folder.path + \"/\" + gps)\n",
    "                            if flip:  # Reorder the data if using CSRS-PPP 2024 pre 2018\n",
    "                                ind_data = ind_data.reindex(index=ind_data.index[::-1])\n",
    "                            data = pd.concat([data, ind_data], ignore_index=True)\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8e16a07f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "la16 /mnt/d/csrs_2024/all/la16\n",
      "la17 /mnt/d/csrs_2024/all/la17\n",
      "la18 /mnt/d/csrs_2024/all/la18\n",
      "slw1 /mnt/d/csrs_2024/all/slw1\n",
      "ws04 /mnt/d/csrs_2024/all/ws04\n",
      "ws05 /mnt/d/csrs_2024/all/ws05\n",
      "gz01 /mnt/d/csrs_2024/all/gz01\n",
      "gz02 /mnt/d/csrs_2024/all/gz02\n",
      "gz03 /mnt/d/csrs_2024/all/gz03\n",
      "gz04 /mnt/d/csrs_2024/all/gz04\n",
      "gz05 /mnt/d/csrs_2024/all/gz05\n",
      "gz06 /mnt/d/csrs_2024/all/gz06\n",
      "gz07 /mnt/d/csrs_2024/all/gz07\n",
      "gz08 /mnt/d/csrs_2024/all/gz08\n",
      "gz09 /mnt/d/csrs_2024/all/gz09\n",
      "gz10 /mnt/d/csrs_2024/all/gz10\n",
      "gz11 /mnt/d/csrs_2024/all/gz11\n",
      "gz12 /mnt/d/csrs_2024/all/gz12\n",
      "gz13 /mnt/d/csrs_2024/all/gz13\n",
      "gz14 /mnt/d/csrs_2024/all/gz14\n",
      "gz15 /mnt/d/csrs_2024/all/gz15\n",
      "gz16 /mnt/d/csrs_2024/all/gz16\n",
      "gz17 /mnt/d/csrs_2024/all/gz17\n",
      "gz18 /mnt/d/csrs_2024/all/gz18\n",
      "gz19 /mnt/d/csrs_2024/all/gz19\n",
      "gz20 /mnt/d/csrs_2024/all/gz20\n",
      "2016\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1290/2913353795.py:38: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  d = pd.read_csv(file, skiprows = 7, delim_whitespace=True)\n",
      "/tmp/ipykernel_1290/2913353795.py:45: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  d = pd.read_csv(file, skiprows = 5, delim_whitespace=True)\n",
      "/tmp/ipykernel_1290/2913353795.py:51: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  d = pd.read_csv(file, skiprows = 6, delim_whitespace=True)\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'LON(d)'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mParserError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 25\u001b[0m, in \u001b[0;36mload\u001b[0;34m(file)\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m     24\u001b[0m     \u001b[38;5;66;03m#CSRS-PPP 2024\u001b[39;00m\n\u001b[0;32m---> 25\u001b[0m     d \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mskiprows\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m3\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m\\\u001b[39;49m\u001b[38;5;124;43ms+\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     26\u001b[0m     data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlongitude\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLONDD\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLONMN\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLONSS\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/gps/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1026\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1024\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m-> 1026\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/gps/lib/python3.11/site-packages/pandas/io/parsers/readers.py:626\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    625\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m parser:\n\u001b[0;32m--> 626\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mparser\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/gps/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1923\u001b[0m, in \u001b[0;36mTextFileReader.read\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m   1917\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1918\u001b[0m     \u001b[38;5;66;03m# error: \"ParserBase\" has no attribute \"read\"\u001b[39;00m\n\u001b[1;32m   1919\u001b[0m     (\n\u001b[1;32m   1920\u001b[0m         index,\n\u001b[1;32m   1921\u001b[0m         columns,\n\u001b[1;32m   1922\u001b[0m         col_dict,\n\u001b[0;32m-> 1923\u001b[0m     ) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# type: ignore[attr-defined]\u001b[39;49;00m\n\u001b[1;32m   1924\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnrows\u001b[49m\n\u001b[1;32m   1925\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1926\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/gps/lib/python3.11/site-packages/pandas/io/parsers/c_parser_wrapper.py:234\u001b[0m, in \u001b[0;36mCParserWrapper.read\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m    233\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlow_memory:\n\u001b[0;32m--> 234\u001b[0m     chunks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_reader\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_low_memory\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    235\u001b[0m     \u001b[38;5;66;03m# destructive to chunks\u001b[39;00m\n",
      "File \u001b[0;32mparsers.pyx:838\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader.read_low_memory\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mparsers.pyx:905\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._read_rows\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mparsers.pyx:874\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._tokenize_rows\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mparsers.pyx:891\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._check_tokenize_status\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mparsers.pyx:2061\u001b[0m, in \u001b[0;36mpandas._libs.parsers.raise_parser_error\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mParserError\u001b[0m: Error tokenizing data. C error: Calling read(nbytes) on source failed. Try engine='python'.",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/gps/lib/python3.11/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'LON(d)'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 39\u001b[0m, in \u001b[0;36mload\u001b[0;34m(file)\u001b[0m\n\u001b[1;32m     38\u001b[0m d \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(file, skiprows \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m7\u001b[39m, delim_whitespace\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m---> 39\u001b[0m data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlongitude\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43md\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mLON(d)\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLON(m)\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLON(s)\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\n\u001b[1;32m     40\u001b[0m data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlatitude\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLAT(d)\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLAT(m)\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLAT(s)\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/gps/lib/python3.11/site-packages/pandas/core/frame.py:4102\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4101\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 4102\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4103\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n",
      "File \u001b[0;32m~/miniconda3/envs/gps/lib/python3.11/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'LON(d)'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/gps/lib/python3.11/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'LONDD'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 46\u001b[0m, in \u001b[0;36mload\u001b[0;34m(file)\u001b[0m\n\u001b[1;32m     45\u001b[0m d \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(file, skiprows \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m5\u001b[39m, delim_whitespace\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m---> 46\u001b[0m data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlongitude\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43md\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mLONDD\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLONMN\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLONSS\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\n\u001b[1;32m     47\u001b[0m data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlatitude\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLATDD\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLATMN\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLATSS\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/gps/lib/python3.11/site-packages/pandas/core/frame.py:4102\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4101\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 4102\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4103\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n",
      "File \u001b[0;32m~/miniconda3/envs/gps/lib/python3.11/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'LONDD'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/gps/lib/python3.11/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'LON(d)'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 13\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;66;03m# Make empty dataframe for each station to be filled in with starts/ends\u001b[39;00m\n\u001b[1;32m     12\u001b[0m out \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(columns\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mx\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124my\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstart\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mend\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstation\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstano\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[0;32m---> 13\u001b[0m stream \u001b[38;5;241m=\u001b[39m \u001b[43mdatastream\u001b[49m\u001b[43m(\u001b[49m\u001b[43msta\u001b[49m\u001b[43m,\u001b[49m\u001b[43myears\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;66;03m#Load data\u001b[39;00m\n\u001b[1;32m     14\u001b[0m x \u001b[38;5;241m=\u001b[39m stream[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mx\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39miloc[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m     15\u001b[0m y \u001b[38;5;241m=\u001b[39m stream[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124my\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39miloc[\u001b[38;5;241m0\u001b[39m]\n",
      "Cell \u001b[0;32mIn[4], line 22\u001b[0m, in \u001b[0;36mdatastream\u001b[0;34m(dir, years)\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m gps \u001b[38;5;129;01min\u001b[39;00m os\u001b[38;5;241m.\u001b[39mlistdir(folder\u001b[38;5;241m.\u001b[39mpath):\n\u001b[1;32m     21\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m gps\u001b[38;5;241m.\u001b[39mendswith(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.pos\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m gps\u001b[38;5;241m.\u001b[39mstartswith(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m'\u001b[39m): \u001b[38;5;66;03m#Ignore xyzt, zip files\u001b[39;00m\n\u001b[0;32m---> 22\u001b[0m         ind_data, flip \u001b[38;5;241m=\u001b[39m \u001b[43mload\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfolder\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m/\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mgps\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     23\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m flip: \u001b[38;5;66;03m#Reorder the data if using CSRS-PPP 2024 pre 2018\u001b[39;00m\n\u001b[1;32m     24\u001b[0m             ind_data \u001b[38;5;241m=\u001b[39m ind_data\u001b[38;5;241m.\u001b[39mreindex(index\u001b[38;5;241m=\u001b[39mind_data\u001b[38;5;241m.\u001b[39mindex[::\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m])\n",
      "Cell \u001b[0;32mIn[7], line 52\u001b[0m, in \u001b[0;36mload\u001b[0;34m(file)\u001b[0m\n\u001b[1;32m     50\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[1;32m     51\u001b[0m     d \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(file, skiprows \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m6\u001b[39m, delim_whitespace\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m---> 52\u001b[0m     data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlongitude\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43md\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mLON(d)\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLON(m)\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLON(s)\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\n\u001b[1;32m     53\u001b[0m     data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlatitude\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLAT(d)\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLAT(m)\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m \u001b[38;5;241m-\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLAT(s)\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m60\u001b[39m\n\u001b[1;32m     54\u001b[0m     data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtime\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mto_datetime(d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mYEAR-MM-DD\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mT\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m d[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHR:MN:SS.SSS\u001b[39m\u001b[38;5;124m'\u001b[39m])\n",
      "File \u001b[0;32m~/miniconda3/envs/gps/lib/python3.11/site-packages/pandas/core/frame.py:4102\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4100\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   4101\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 4102\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4103\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   4104\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/miniconda3/envs/gps/lib/python3.11/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3810\u001b[0m     ):\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'LON(d)'"
     ]
    }
   ],
   "source": [
    "dir = \"/mnt/d/csrs_2024/all\"\n",
    "\n",
    "\n",
    "stano = 0\n",
    "for sta in os.scandir(dir):\n",
    "    if sta.is_dir():\n",
    "        stano = stano + 1\n",
    "        print(sta.name, sta.path)\n",
    "        if sta.name == \"gz20\":\n",
    "            years = [\n",
    "                \"2006\",\n",
    "                \"2007\",\n",
    "                \"2008\",\n",
    "                \"2009\",\n",
    "                \"2010\",\n",
    "                \"2011\",\n",
    "                \"2012\",\n",
    "                \"2013\",\n",
    "                \"2014\",\n",
    "                \"2015\",\n",
    "                \"2016\",\n",
    "                \"2017\",\n",
    "                \"2018\",\n",
    "                \"2019\",\n",
    "                \"2020\",\n",
    "            ]\n",
    "            # Make empty dataframe for each station to be filled in with starts/ends\n",
    "            out = pd.DataFrame(columns=[\"x\", \"y\", \"start\", \"end\", \"station\", \"stano\"])\n",
    "            stream = datastream(sta, years)  # Load data\n",
    "            x = stream[\"x\"].iloc[0]\n",
    "            y = stream[\"y\"].iloc[0]\n",
    "            # print(stream)\n",
    "            start_time = stream[\"time\"][0]\n",
    "            for i in range(len(stream[\"time\"])):\n",
    "                if i > 0:\n",
    "                    prior_time = stream[\"time\"][i - 1]\n",
    "                    current_time = stream[\"time\"][i]\n",
    "\n",
    "                    delta_t = current_time - prior_time\n",
    "                    if delta_t > datetime.timedelta(\n",
    "                        days=1, minutes=-1\n",
    "                    ):  # Check if > 1 day\n",
    "                        end_time = prior_time\n",
    "                        # print(start_time)\n",
    "                        # print(end_time)\n",
    "                        out.loc[len(out.index)] = [\n",
    "                            x,\n",
    "                            y,\n",
    "                            start_time,\n",
    "                            end_time,\n",
    "                            sta.name,\n",
    "                            stano,\n",
    "                        ]\n",
    "                        start_time = current_time\n",
    "\n",
    "                    if i == len(stream[\"time\"]) - 1:\n",
    "                        end_time = current_time\n",
    "\n",
    "                        out.loc[len(out.index)] = [\n",
    "                            x,\n",
    "                            y,\n",
    "                            start_time,\n",
    "                            end_time,\n",
    "                            sta.name,\n",
    "                            stano,\n",
    "                        ]\n",
    "\n",
    "            print(out)\n",
    "\n",
    "            out[\"start\"] = out[\"start\"].dt.strftime(\"%Y-%m-%dT%H:%M:%S.%f\")\n",
    "            out[\"end\"] = out[\"end\"].dt.strftime(\"%Y-%m-%dT%H:%M:%S.%f\")\n",
    "            out.to_csv(\n",
    "                \"sta_uptime_plot.txt\",\n",
    "                sep=\"\\t\",\n",
    "                index=False,\n",
    "                mode=\"a\",\n",
    "                header=not \"sta_uptime_plot.txt\",\n",
    "            )\n",
    "\n",
    "            with open(\"sta_uptime_plot.txt\", \"a\") as file:\n",
    "                file.write(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9de2bdc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                               0                           1     2   3\n",
      "0     2010-01-15T00:40:00.000000  2010-05-02T15:59:00.000000  la01   1\n",
      "1     2010-11-09T20:18:45.000000  2011-05-19T22:59:30.000000  la01   1\n",
      "2     2011-11-15T06:44:45.000000  2012-05-16T19:56:45.000000  la01   1\n",
      "3     2012-09-08T23:02:30.000000  2012-09-09T01:45:00.000000  la01   1\n",
      "4     2012-09-10T20:02:15.000000  2012-09-10T22:08:00.000000  la01   1\n",
      "...                          ...                         ...   ...  ..\n",
      "1329  2016-09-03T09:20:45.000000  2016-09-03T21:25:00.000000  mg04  45\n",
      "1330  2016-09-05T21:14:30.000000  2016-11-26T20:57:00.000000  mg04  45\n",
      "1331  2014-01-24T02:34:00.000000  2014-12-17T23:59:45.000000  mg05  46\n",
      "1332  2014-01-23T22:39:00.000000  2014-12-18T03:15:30.000000  mg06  47\n",
      "1333  2014-01-27T01:18:45.000000  2014-12-18T23:59:45.000000  mg07  48\n",
      "\n",
      "[1334 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "# Load File and get true times with no data\n",
    "\n",
    "times = pd.read_csv(\"sta_uptime_input.txt\", sep=\"\\t\", header=None)\n",
    "print(times)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "535f5da2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
